vopr: search for correlated faults after safety mode

chaitanyabhandari opened the pull request:
This PR solves the following two VOPR failures:

#### ./zig/zig build -Drelease vopr -- 9161948212551073841

In this one, an uncommitted prepare is missing on more than a nack quorum of replicas (it is corrupted on R2, was never replicated to R0, and R1 goes into recovering_head state). Specifically, the cluster ends up getting stuck during view change, and all replicas continuously `on_do_view_change: view=22334 quorum received, awaiting repair`.

This is detected as a liveness failure at the end of safety mode, after which we expect replicas to reply to all requests and upgrade to the latest available release. We expect this because we are very careful about the parts of the grid & WAL where we inject corruptions, and we also take special measures to avoid more than a view change quorum of replicas to go into recovering_head status. Future work (https://github.com/tigerbeetle/tigerbeetle/pull/2708) will get rid of all of this logic that cautiously does storage fault injection.

For now, we roll out part of https://github.com/tigerbeetle/tigerbeetle/pull/2708, specifically the logic that tries to detect correlated faults at the end of safety mode as well (similar to how we do at the end of liveness mode). This would help shake off such false positives, where we accidentally end up injecting more faults than we should during safety mode.  Also, we add logic to the StateChecker to track uncommitted headers, so we can check for their existence in the VOPR.

#### ./zig/zig build -Drelease vopr -- 13232551907466493063

We end up violating one of our invariants at the end of liveness mode here -- a grid block that can be repaired (is present on the core) should be repaired.

Specifically, a replica that is supposed to help out with a block keeps emitting `on_request_blocks: ignoring remaining blocks; busy`. This happened because _no asynchronous callbacks were being invoked_, causing all grid repair read IOPs to get stuck. The ultimate root cause was that paused replicas weren't being unpaused while transitioning to liveness mode, leading to storage not being ticked!

chaitanyabhandari committed:
vopr: unpause paused replicas while entering liveness mode

chaitanyabhandari committed:
vopr: fix count for correlated block corruptions in the cluster

chaitanyabhandari committed:
vopr: simplify logic for finding correlated missing prepares

chaitanyabhandari committed:
vopr: search for correlated faults after safety mode

sentientwaffle commented on src/vopr.zig:
```diff
@@ -624,11 +636,6 @@ pub const Simulator = struct {
     /// See https://tigerbeetle.com/blog/2023-07-06-simulation-testing-for-liveness for broader
     /// context.
     pub fn transition_to_liveness_mode(simulator: *Simulator) void {
```
Since the correct way to use this function is to set `simulator.core` immediately beforehand, how about we instead pass it the `core` bitset as an argument?

sentientwaffle commented on src/vopr.zig:
```diff
@@ -637,6 +644,12 @@ pub const Simulator = struct {
                 simulator.replica_restart(@intCast(replica_index), fault);
             }
+
+            const replica_health = simulator.cluster.replica_health[replica_index];
+            if (replica_health == .up and replica_health.up.paused) {
+                simulator.cluster.replica_unpause(@intCast(replica_index));
+            }
```
This bug is a great find!
(I am also surprised it took us this long to hit a seed that got stuck this way.)

sentientwaffle commented on src/vopr.zig:
```diff
@@ -696,117 +709,104 @@ pub const Simulator = struct {
+                if (!simulator.core.isSet(replica.replica) or
+                    !replica.journal.has_prepare(&header))
+                {
+                    replicas_missing_ops[index].set(replica.replica);
```
```suggestion
                    replicas_missing_op.set(replica.replica);
```

sentientwaffle commented on src/testing/cluster/state_checker.zig:
```diff
@@ -255,10 +269,18 @@ pub fn StateCheckerType(comptime Client: type, comptime Replica: type) type {
+                    break :blk commit.header;
+                } else {
+                    const slot = op % constants.pipeline_prepare_queue_max;
+                    break :blk state_checker.uncommitted_headers[slot];
```
How do we know that this is the uncommitted prepare that the cluster is stuck on?
e.g. if we prepare `(view=1, op=10)` and then prepare `(view=2, op=10)`, the latter is in `uncommitted_headers`, but a `view=3` might select the former.

chaitanyabhandari replied:
This version of the code was indeed vulnerable to this. It was solved via https://github.com/tigerbeetle/tigerbeetle/pull/2797/commits/20710a086c6af03eaac0ce4fa0b66f9bc840cb9c.

chaitanyabhandari committed:
vopr: address review comments

sentientwaffle commented on src/testing/cluster/state_checker.zig:
```diff
@@ -93,6 +101,12 @@ pub fn StateCheckerType(comptime Client: type, comptime Replica: type) type {
         }
 
+        pub fn on_new_prepare(state_checker: *StateChecker, prepare: *const Message.Prepare) void {
+            assert(prepare.header.command == .prepare);
```
Can we also assert that the op number does not decrease in a given slot?

sentientwaffle replied:
Also can we assert that the op is not more than pipeline max ahead of the currently highest committed op.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -363,36 +363,41 @@ pub fn main() !void {
     } else {
+        if (cli_args.lite) return;
+
+        // Heal current network partitions, disable future process, storage, and network faults
```
Maybe add one more sentence to the start of this comment for context. Something along the lines of "safety mode ran out of ticks without completing its requests, so now we check whether it was correct to do so".

sentientwaffle commented on src/vopr.zig:
```diff
@@ -363,36 +363,41 @@ pub fn main() !void {
+        simulator.options.cluster.standby_count,
+    );
+
+    simulator.transition_to_liveness_mode(simulator.core);
```
I was imagining that we could create the core with `const core = ...` then `transition_to_liveness_mode` would be responsible for assigning it to `simulator.core`. (And also it could assert the old core state?)

chaitanyabhandari replied:
Ah, okay! I misunderstood the intention. Thanks for clarifying.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -715,117 +728,104 @@ pub const Simulator = struct {
-                    }
+        for (replicas_missing_ops, 0..) |*replicas_missing_op, index| {
+            replicas_missing_op.* = ReplicaSet.initEmpty();
+            const op = index + cluster_op_repair_min;
```
Instead of starting the iteration index from 0 could we start it from `cluster_op_repair_min`? Then we would be iterating the ops directly.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -835,26 +835,29 @@ pub const Simulator = struct {
-            address: u64,
-            checksum: u128,
-        }).init(allocator);
+        const FaultyReplicas = std.StaticBitSet(constants.replicas_max + constants.standbys_max);
```
```suggestion
        const FaultyReplicas = std.StaticBitSet(constants.members_max);
```

sentientwaffle commented on src/vopr.zig:
```diff
@@ -912,10 +920,10 @@ pub const Simulator = struct {
         }
 
-        if (blocks_missing.items.len == 0) {
+        if (blocks_missing.count() == 0) {
```
Ah, so before if the same block was missing on multiple replicas then we would count it as missing multiple times.

chaitanyabhandari committed:
vopr: infer uncommitted headers during liveness checks

sentientwaffle commented on src/testing/cluster/state_checker.zig:
```diff
@@ -65,6 +65,12 @@ pub fn StateCheckerType(comptime Client: type, comptime Replica: type) type {
+            var uncommitted_headers: [pipeline_max]vsr.Header.Prepare = undefined;
+            for (0..pipeline_max) |slot| {
+                uncommitted_headers[slot] = vsr.Header.Prepare.reserve(options.cluster_id, slot);
+            }
```
This code can be removed.

chaitanyabhandari replied:
Oops, nice catch! For this PR, I was trying to contain all related changes within a single commit but oh well... 😄

sentientwaffle commented on src/testing/cluster/state_checker.zig:
```diff
@@ -254,7 +260,8 @@ pub fn StateCheckerType(comptime Client: type, comptime Replica: type) type {
         }
 
-        pub fn header_with_op(state_checker: *StateChecker, op: u64) vsr.Header.Prepare {
+        pub fn committed_header_with_op(state_checker: *StateChecker, op: u64) vsr.Header.Prepare {
```
Since we are back to all of the state checkers headers being committed we can move that from the name.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -715,18 +737,26 @@ pub const Simulator = struct {
+            .recovering_head => return false,
+            .view_change => {
+                // Lagging replicas do not initiate WAL repair during view change.
+                if (vsr.Checkpoint.durable(replica.op_checkpoint_next(), replica.commit_max)) {
```
This could instead simply be `return !vsr.Checkpoint.durable(replica.op_checkpoint_next(), replica.commit_max)`.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -736,96 +766,117 @@ pub const Simulator = struct {
+                op_repair_min = replica.op_repair_min();
+            } else {
+                op_repair_min = @min(op_repair_min.?, replica.op_repair_min());
             }
```
If we initialize `op_repair_min` to `std.math.maxInt(u16)` then we can simplify this and avoid the branching.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -736,96 +766,117 @@ pub const Simulator = struct {
+        var cluster_op_head: u64 = 0;
+        var cluster_commit_max: u64 = 0;
+        var cluster_log_view: u32 = 0;
+        var op_repair_min: ?u64 = 0;
```
I think this is a bug -- by initializing this to zero, the minimum will always be zero.

chaitanyabhandari replied:
Ah, I wanted this to be null! Great catch. 😶‍🌫️ 

Anyway, https://github.com/tigerbeetle/tigerbeetle/pull/2797#discussion_r1995775096 would obviate the need for this!

chaitanyabhandari committed:
vopr: handle correlated corruptions in root prepare

sentientwaffle commented on src/vopr.zig:
```diff
@@ -736,96 +766,117 @@ pub const Simulator = struct {
+        // smaller log_view may have an outdated version of the same uncommitted op. There may be
+        // at most a pipeline of uncommitted headers in the cluster.
+        const pipeline_max = constants.pipeline_prepare_queue_max;
+        var uncommitted_headers: [pipeline_max]vsr.Header.Prepare = undefined;
```
Maybe use an optional type (accessed later with `.?`) for the array entries just to be careful.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -736,96 +766,117 @@ pub const Simulator = struct {
+            } else {
+                op_repair_min = @min(op_repair_min.?, replica.op_repair_min());
             }
         }
```
Can we assert that the `cluster_commit_max` is less than or equal to the `cluster_op_head`.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -736,96 +766,117 @@ pub const Simulator = struct {
-                            if (missing_prepare_op == null or op < missing_prepare_op.?) {
-                                missing_prepare_op = op;
-                            }
+                if (cluster_commit_max < cluster_op_head) {
```
We already checked this condition above.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -736,96 +766,117 @@ pub const Simulator = struct {
+        // Check whether any of the uncommitted headers is corrupted on more than a nack
+        // quorum of replicas. If so, the cluster cannot initiate repair or commit (see the
+        // awaiting_repair and complete_invalid cases in the DVCQuorum).
+        if (cluster_commit_max <= cluster_op_head) {
```
Hmm, I mentioned above that I thought this could be an assert, but maybe it can't be? When is `cluster_commit_max <= cluster_op_head` not true?

chaitanyabhandari replied:
Mhmmmmm, I actually think that it _can_ be an asserted, IIUC we always update commit_max & op_head in lockstep (via `set_op_and_commit_max`)? This was more an unconscious condition that I added, because I was adding it for `replica.commit_min < replica.commit_max` and `replica.op_repair_min() <= replica.commit_min`.

I'll add it!

chaitanyabhandari commented:
Thanks for the great comments @sentientwaffle, I'm much happier with how the changes look now! I've addressed the comments in ca06d7517e9007814ca06e51dba791f10045cae1.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -737,96 +766,125 @@ pub const Simulator = struct {
+            const header = blk: {
+                if (op > cluster_commit_max) {
+                    const uncommitted_header = uncommitted_headers[op % pipeline_max].?;
+                    assert(uncommitted_header.op > cluster_commit_max);
```
Ah actually since we just checked `op > cluster_commit_max` we could check `uncommitted_header.op == op`.

chaitanyabhandari replied:
Ack, done in https://github.com/tigerbeetle/tigerbeetle/pull/2797/commits/485051fa5a0608fbb299552a7e1b1fc73072a913.

sentientwaffle commented on src/vopr.zig:
```diff
@@ -737,96 +766,125 @@ pub const Simulator = struct {
+                const header = blk: {
+                    if (op > cluster_commit_max) {
+                        const uncommitted_header = uncommitted_headers[op % pipeline_max].?;
+                        assert(uncommitted_header.op > cluster_commit_max);
```
Same as https://github.com/tigerbeetle/tigerbeetle/pull/2797/files#r1996043454

chaitanyabhandari committed:
vopr: refactors, better asserts around uncommitted_headers
